server/events/vcs/gitea/client.go

```diff
@@ -0,0 +1,468 @@
+// Copyright 2024 Martijn van der Kleijn & Florian Beisel
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gitea
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"strings"
+	"time"
+
+	"code.gitea.io/sdk/gitea"
+	"github.com/pkg/errors"
+	"github.com/runatlantis/atlantis/server/events/models"
+	"github.com/runatlantis/atlantis/server/logging"
+)
+
+type GiteaClient struct {
+	giteaClient *gitea.Client
+	username    string
+	token       string
+	ctx         context.Context
+	logger      logging.SimpleLogging
+}
+
+type GiteaPRReviewSummary struct {
+	Reviews []GiteaReview
+}
+
+type GiteaReview struct {
+	ID          int64
+	Body        string
+	Reviewer    string
+	State       gitea.ReviewStateType // e.g., "APPROVED", "PENDING", "REQUEST_CHANGES"
+	SubmittedAt time.Time
+}
+
+type GiteaPullGetter interface {
+	GetPullRequest(repo models.Repo, pullNum int) (*gitea.PullRequest, error)
+}
+
+// NewClient builds a client that makes API calls to Gitea. httpClient is the
+// client to use to make the requests, username and password are used as basic
+// auth in the requests, baseURL is the API's baseURL, ex. https://corp.com:7990.
+// Don't include the API version, ex. '/1.0'.
+func NewClient(baseURL string, username string, token string, logger logging.SimpleLogging) (*GiteaClient, error) {
+	giteaClient, err := gitea.NewClient(baseURL,
+		gitea.SetToken(token),
+		gitea.SetUserAgent("atlantis"),
+	)
+
+	if err != nil {
+		return nil, errors.Wrap(err, "creating gitea client")
+	}
+
+	return &GiteaClient{
+		giteaClient: giteaClient,
+		username:    username,
+		token:       token,
+		ctx:         context.Background(),
+		logger:      logger,
+	}, nil
+}
+
+func (c *GiteaClient) GetPullRequest(repo models.Repo, pullNum int) (*gitea.PullRequest, error) {
+	pr, _, err := c.giteaClient.GetPullRequest(repo.Owner, repo.Name, int64(pullNum))
+
+	if err != nil {
+		return nil, err
+	}
+
+	return pr, nil
+}
+
+// GetModifiedFiles returns the names of files that were modified in the merge request
+// relative to the repo root, e.g. parent/child/file.txt.
+func (c *GiteaClient) GetModifiedFiles(repo models.Repo, pull models.PullRequest) ([]string, error) {
+	changedFiles := make([]string, 0)
+	page := 0
+	nextPage := 1
+	listOptions := gitea.ListPullRequestFilesOptions{
+		ListOptions: gitea.ListOptions{
+			Page:     1,
+			PageSize: 100,
```

Comment:
Is this the default pagesize?

Do we want this option to be configurable?

According to https://pkg.go.dev/code.gitea.io/sdk/gitea#ListOptions, it seems that it's configurable on the gitea server so some users may need to increase or decrease it.

Reply:
You're right, this should probably be configurable. I'll add a flag to allow for that. The value is dependent on the specific Gitea installation's configuration. I'll apply the default from the Gitea docs.

Reply:
Added --gitea-page-size, defaulting to 30 based on Gitea docs.